### PR TITLE
KAFKA-12961; Verify group generation in `DelayedJoin`

### PR DESCRIPTION
**Pull request by dajac**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala [hunk 1]
```diff
@@ -1391,78 +1392,92 @@ class GroupCoordinator(val brokerId: Int,
...
-          Seq(GroupJoinKey(group.groupId)))
+      if (generationId != group.generationId) {
+        error(s"Received unexpected notification of join complete for ${group.groupId} " +
+          s"with an old generation $generationId while the group has ${group.generationId}.")
       } else {
-        group.initNextGeneration()
-        if (group.is(Empty)) {
-          info(s"Group ${group.groupId} with generation ${group.generationId} is now empty " +
-            s"(${Topic.GROUP_METADATA_TOPIC_NAME}-${partitionFor(group.groupId)})")
+        val notYetRejoinedDynamicMembers = group.notYetRejoinedMembers.filterNot(_._2.isStaticMember)
```
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:1418]: It looks like this branch is all of the existing logic, but thought I'd check if there are any changes to look at.

**Review by hachikuji — COMMENTED** [review]
Thanks for the PR. It makes sense to me. Would it be possible to add a test for the case when a late `DelayedJoin` completes?

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala (hunk 1 above)
Comment (dajac) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:1418]: That's correct.

**Comment by dajac** [comment]
> Thanks for the PR. It makes sense to me. Would it be possible to add a test for the case when a late `DelayedJoin` completes?

Yeah, let me see if I can add something.

**Comment by ijuma** [comment]
@dajac is this still relevant?

**Pull request closed without merging**